### PR TITLE
Fix MAP metric for empty cases

### PR DESCRIPTION
**Pull request by tkupek**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/metrics/issues/601 (empty tensors in metric)
Fixes https://github.com/PyTorchLightning/metrics/issues/626 (DDP stuck in multi GPU mode)
Fixes an issue if the metric is evaluated with empty lists

Tests where added for both cases

## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: torchmetrics/detection/map.py [hunk 1]
```diff
@@ -405,8 +408,11 @@ def _get_coco_format(
...
                 annotation_id += 1
 
-        classes = [
-            {"id": i, "name": str(i)}
-            for i in torch.cat(self.detection_labels + self.groundtruth_labels).unique().cpu().tolist()
-        ]
+        classes = [{"id": i, "name": str(i)} for i in self._get_classes()]
         return {"images": images, "annotations": annotations, "categories": classes}
+
+    def _get_classes(self):
```
Comment (Borda) [inline torchmetrics/detection/map.py:414]:
```suggestion
    def _get_classes(self) -> list:
```

**Review by Borda — APPROVED** [review]

**Comment by SkafteNicki** [comment]
@tkupek How is it going here?

**Review by SkafteNicki — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
